beets/test/fixtures.py

```diff
@@ -0,0 +1,92 @@
+# This file is part of beets.
+# Copyright 2024, Arav K.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""
+This module provides `pytest`-based fixtures for testing Beets.
+"""
+
+from pathlib import Path
+from typing import Iterator
+
+import pytest
+from confuse import Configuration
+
+import beets
+from beets.library import Library
+
+
+@pytest.fixture(scope="session")
+def resource_dir() -> Path:
+    """
+    A fixture returning the resource directory for tests.
+    """
+
+    return Path(__file__).parents[2] / "test" / "rsrc"
+
+
+@pytest.fixture(scope="function")
+def config(
+    tmp_path: Path,
+    monkeypatch: pytest.MonkeyPatch,
+) -> Iterator[Configuration]:
+    """
+    A fixture for Beets configuration.
+    """
+
+    # 'confuse' looks at `HOME`, so we set it to a tmpdir.
+    monkeypatch.setenv("HOME", str(tmp_path))
+
+    # Initialize configuration.
+    config = beets.config
+    config.sources = []
+    config.read(user=False, defaults=True)
+
+    # Set configuration defaults.
+    config["plugins"] = []
+    config["verbose"] = 1
+    config["ui"]["color"] = False
+    config["threaded"] = False
+    config["directory"] = str(tmp_path)
+
+    # Provide the configuration to the test.
+    yield config
+
+    # Reset the global configuration state.
+    beets.config.clear()
+    beets.config._materialized = False
+
+
+@pytest.fixture(scope="function")
+def library(
+    tmp_path: Path,
+    config: Configuration,
```

Comment:
`config` fixture is not used within this fixture.

If it was placed here to ensure it runs, consider [making it an autouse fixture](https://docs.pytest.org/en/stable/how-to/fixtures.html#autouse-fixtures-fixtures-you-don-t-have-to-request).


```python
@pytest.fixture(autouse=True)
def config(...)
```

alternatively, if you only need the config setup in some specific tests

```python
# note leading underscore which is pytest convention for defining a fixture that doesn't return anything
@pytest.fixture
def _config(...):
    ...


@pytest.mark.usefixtures("_config")
def test_something(...):
    ...
```

or use it across the entire module
```python
# test/test_something.py
pytestmark = pytest.mark.usefixtures("_config")

...

def test_something(...):
    ...
```

Reply:
This is interesting.  I based this fixture on the way the existing testing infrastructure works.  When setting up a `BeetsTestCase`, the line `self.config = beets.config` occurs; this means that `self.config` is basically an alias for the same global variable.  I think the motivation behind this was that the Beets API could one day be re-entrant, relying on a `config` object passed into the `Library` rather than just the global `beets.config`.  This is why the `config` fixture returns a reference to `beets.config`: it is the direct replacement for `self.config` in `BeetsTestCase`.

While the `library` fixture does not _directly_ rely on the `config` fixture, it does rely on the `beets.config` global variable to be reset at the start of every test.  On the other hand, I don't think the `config` fixture should be auto-use; at least for now, I want users to explicitly mention it, so we can gain some understanding of exactly how it is used across the tests.  For the same reason, I don't think it should be renamed to `_config`: the fact that we don't directly reference it in `library()` is more of an implementation detail, and it won't be true for other users of the fixture.

Reply:
> For the same reason, I don't think it should be renamed to `_config`: the fact that we don't directly reference it in `library()` is more of an implementation detail, and it won't be true for other users of the fixture.

I agree - at the end of the day it __does__ return the configuration object.

The unused `config` argument makes some of the linters screech (for example `flake8-unused-arguments`). You can add `pytest.mark.usefixtures("config")` above this fixture definition. I like `usefixtures` because it communicates that the `library` fixture __only__ requires the `config` fixture set up - it does not care what `config` returns.

Reply:
Alright, will do.  Thanks!